lib/active_model/serializer/adapter/fragment_cache.rb

```diff
@@ -82,30 +81,36 @@ def cached_attributes(klass, serializers)
         #     User_AdminCachedSerializer
         #     User_AdminNOnCachedSerializer
         #
-        def fragment_serializer(name, klass)
+        def fragment_serializer(name)
+          klass      = serializer.class
           cached     = "#{to_valid_const_name(name)}CachedSerializer"
           non_cached = "#{to_valid_const_name(name)}NonCachedSerializer"
 
-          Object.const_set cached, Class.new(ActiveModel::Serializer) unless Object.const_defined?(cached)
-          Object.const_set non_cached, Class.new(ActiveModel::Serializer) unless Object.const_defined?(non_cached)
+          cached_serializer     = get_or_create_serializer(cached)
+          non_cached_serializer = get_or_create_serializer(non_cached)
 
           klass._cache_options ||= {}
-          klass._cache_options[:key] = klass._cache_key if klass._cache_key
-
-          cached.constantize.cache(klass._cache_options)
+          cache_key = klass._cache_key
+          klass._cache_options[:key] = cache_key if cache_key
+          cached_serializer.cache(klass._cache_options)
 
-          # Preserve the type setting in the cached/non-cached serializer classes
-          cached.constantize.type(klass._type)
-          non_cached.constantize.type(klass._type)
+          type = klass._type
+          cached_serializer.type(type)
```

Comment:
I was not able to refactor the next few lines as there were some specs failing which I assume was due the issue explained [here](https://github.com/rails-api/active_model_serializers/issues/1520#issuecomment-186397391).